src/assignment-10/FlashLoanLiquidator.sol

```diff
@@ -0,0 +1,123 @@
+pragma solidity ^0.8.13;
+
+import { IERC20 } from "yield-utils-v2/token/IERC20.sol";
+import { IUniswapV2Pair } from "src/assignment-10/interfaces/IUniswapV2Pair.sol";
+import { IUniswapV2Factory } from "src/assignment-10/interfaces/IUniswapV2Factory.sol";
+import { CollateralizedVault } from "src/assignment-10/CollateralizedVault.sol";
+import { AMM } from "src/assignment-7/AMM.sol";
+import { TransferHelper } from "yield-utils-v2/token/TransferHelper.sol";
+
+/// @title Flash loan liquidator
+/// @author davidbrai
+/// @notice Uses a flash loan in order to liquidate an undercollateralized debt position and keep the arbitrage
+/// @dev Uses UniswapV2Pair for flash swaps
+contract FlashLoanLiquidator {
+
+    using TransferHelper for IERC20;
+
+    /******************
+     * Immutables
+     ******************/
+
+    IERC20 immutable public underlying;
+    IERC20 immutable public collateral;
+    IUniswapV2Factory immutable public uniswapFactory;
+    CollateralizedVault immutable public vault;
+    AMM immutable public amm;
+
+
+    /******************
+     * Storage
+     ******************/
+
+    /// @dev Security: this is used to make sure the callback is only called by the expected UniV2 pair contract
+    address public permissionedPair;
+
+    /******************
+     * Events
+     ******************/
+    event Liquidate(address indexed liquidator, address indexed liquidatee, uint256 profit);
+
+    /******************
+     * Errors
+     ******************/
+
+    error UnauthorizedInitiator();
+    error UnauthorizedMsgSender();
+
+    struct SwapParams {
+        address vaultUser;
+        address liquidator;
+    }
+
+    /// @notice Creates a new FlashLoanLiquidator
+    /// @param underlying_ ERC20 token address of the asset being borrowed
+    /// @param collateral_ ERC20 token address of the collateral asset
+    /// @param uniswapFactory_ address of UniswapV2Factory, used for finding UniswapV2Pair contracts to use for flash swaps
+    /// @param vault_ A CollateralizedVault contract against which to liquidate
+    /// @param amm_ An AMM for swapping the `collateral_` back into `underlying_`
+    constructor(address underlying_, address collateral_, address uniswapFactory_, address vault_, address amm_) {
+        underlying = IERC20(underlying_);
+        collateral = IERC20(collateral_);
+        uniswapFactory = IUniswapV2Factory(uniswapFactory_);
+        vault = CollateralizedVault(vault_);
+        amm = AMM(amm_);
+    }
+
+    /// @notice Takes a flash loan of `underlying` and liquidates an undercollateralized user debt position
+    ///     Then is swaps the `collateral` back into `underlying` tokens, repays the flash loan, and send
+    ///     the profit back to who invoked this function.
+    /// @dev UniswapV2Pair is used for flash loaning `underlying`
+    /// @param user The debt position to liquidate in `vault`
+    function liquidate(address user) public {
+        uint256 debt = vault.borrows(user);
```

Comment:
It would be a good idea to check here that the vault is undercollateralized, to avoid wasting a lot of gas if it recollateralized by the time your transaction is mined.

Reply:
added here: https://github.com/davidbrai/yield-mentorship/pull/9/commits/1e65c893bfab4238035f55711829c917b13345fd

Question: if this is sent as a bundle via flashbots, then unlikely that the transaction will be mined if the position is not longer undercollateralized, right?
in that case it actually may be better to remove this check to save on gas for the optimistic case

Reply:
If I'm not wrong flashbots doesn't guarantee that you will get in a particular block, only that when you are in a block you won't get front-run. I need to read on that.

The `liquidate` function will be called from outside the chain, according to the chain conditions for a particular block. By the time the transaction is mined the conditions might have changed.

However, if flashbots guarantees you that you get in the next block no matter what, then yes, the condition would be unnecessary.

Reply:
IIUC flashbots won't include a reverting transaction:

> A reverting transaction (unless specified via [optional argument revertingTxHashes](https://docs.flashbots.net/flashbots-auction/miners/mev-relay/#eth_sendbundle) or uncle'd)

from: https://docs.flashbots.net/flashbots-auction/searchers/advanced/troubleshooting#does-your-transaction-work-and-pay-enough